working/macros/feature-specification.md

```diff
@@ -246,6 +246,41 @@ introspection APIs.
 These macros can fully introspect on any type reachable from the declarations
 they annotate, including introspecting on members of classes, etc.
 
+## Macro Instantiation
+
+Macros should be able to be able to be instantiated using only the libraries
+that are available to the macro definition library. This allows us to compile an
+app (or spawn an isolate) that can run macro applications for that macro using
+only its own dependencies.
+
+### Macro Constructor Parameters
+
+In order to facilitate this, parameters to Macro constructors are limited to
+a specific set of allowable types:
+
+- [Code][] or its subtypes
+- String
+- int, double, num
+- bool
+- null (really this just allows nullable types)
+- Uri
+- DateTime
+
+### Macro Constructor Arguments
+
```

Comment:
This is a little unclear to me about whether arguments are values or meta-values. How about something kind of like:

---

Arguments to macros are passed as *unevaluated code*. In other words, if you apply a macro like:

```dart
@myMacro(1 + 2)
class SomeClass {}
```

The argument passed is not the number 3, but a Code object representing the expression `1 + 2`. This is useful if a macro wants to take an arbitrary piece of syntax as an argument that it then injects into the generated code.

In many cases, though, what the macro author wants is an argument *value*. For example:

```dart
@coordinates(dimensions: 3)
class Vec3 {}
```

Here, the `@coordinates` macro generates, `x`, `y`, etc. fields based on the given dimension passed to the macro. To support this, a macro constructor can accept parameters of primitive types and it will receive the arguments as values, not Code objects representing the value. However, the argument expression must be a *literal* of that type, and not any arbitrary expression that happens to evaluate to that type.

For example:

```dart
@coordinates(dimensions: 1 + 2) // Error.
class Vec3 {}
```

?

Reply:
If the arguments are recorded as expressions, then can you use variables references?
Say:
```dart
const string = "something or other";
...
@MyMacro(string, 42)
...
```
How about interpolations:
```dart
@MyMacro("Header:$string:Trailer")
...
```

If you can't access const variables, then you also can't use enums, which is likely going to affect the APIs that you'd want to build around the macro annotation. 

(How about just serializing the macro annotation constant, deserializing it in the isolate running the macro, then failing compilation if you can't do both).

Reply:
I took another shot at rewording this also. Borrowed some stuff from @munificent, but tried not to mislead people by oversimplifying things early on still.... I think its important to make the pass by value vs Code object distinction early on.